api/common/sharing/utils.go

```diff
@@ -12,64 +12,236 @@ import (
+	 */
+	if len(parameters) == 0 {
+		return util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+	ID, IDExists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	name, nameExists := parameters[instance_sharing.ReferenceInstanceNameSelector]
+	plan, planExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	labels, labelsExists := parameters[instance_sharing.ReferenceLabelSelector]
+	var selectorLabels types.Labels
+	if err := util.BytesToObject([]byte(labels.Raw), &selectorLabels); labelsExists && err != nil {
+		return err
+	}
+	hasAtLeastOneSelector := selectorLabels != nil && len(selectorLabels) > 0 || nameExists && len(name.String()) > 0 || planExists && len(plan.String()) > 0
+	selectorsAndID := IDExists && len(ID.String()) > 0 && hasAtLeastOneSelector
+	emptyValues := len(ID.String()) == 0 && !hasAtLeastOneSelector
+	if selectorsAndID || emptyValues {
+		return util.HandleInstanceSharingError(util.ErrInvalidReferenceSelectors, "")
 	}
+	return nil
+}
+
+func getSharedInstances(ctx context.Context, repository storage.Repository, offeringID, tenantIdentifier, tenantID string) (types.ObjectList, error) {
+	queryParams := map[string]interface{}{
+		"tenant_identifier": tenantIdentifier,
+		"tenant_id":         tenantID,
+		"offering_id":       offeringID,
+	}
+	objectList, err := repository.QueryForList(ctx, types.ServiceInstanceType, storage.QueryForSharedInstances, queryParams)
+	return objectList, err
+}
+
+func filterInstancesBySelectors(ctx context.Context, repository storage.Repository, instances types.ObjectList, parameters map[string]gjson.Result, smaap bool, offeringID string) (*types.ServiceInstance, error) {
+	var err error
+	var filteredInstances types.ServiceInstances
 
-	if !referencedInstance.IsShared() {
-		log.C(ctx).Debugf("The target instance %s is not shared.", referencedInstance.ID)
-		return "", util.HandleInstanceSharingError(util.ErrReferencedInstanceNotShared, referencedInstance.ID)
+	planSelectorVal, planSelectorExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	var selectorPlan *types.ServicePlan
+	if planSelectorExists && len(planSelectorVal.String()) > 0 {
+		selectorPlan, err = getSelectorPlan(ctx, repository, smaap, offeringID, planSelectorVal.String())
+		if err != nil {
+			log.C(ctx).Errorf("Failed to retrieve the plan %s with the offering %s via the selector provided: %s", planSelectorVal.String(), offeringID, err)
+			return nil, err
+		}
 	}
 
-	return referencedInstanceID.String(), nil
+	for i := 0; i < instances.Len(); i++ {
+		// "*" for any shared instance of the tenant, if parameter not passed, set true for any shared instance
+		selectorVal, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+		if exists && len(selectorVal.String()) > 0 && selectorVal.String() != "*" {
+			continue
+		}
+
```

Comment:
this case can be removed
if you had id you won't be in filterInstancesBySelectors func
and you block id with other selectors